src/aliceVision/image/ImageCache.hpp

```diff
@@ -0,0 +1,417 @@
+{
+    if (downscaleLevel < 1)
+    {
+        ALICEVISION_THROW_ERROR("[image] ImageCache: cannot load image with downscale level < 1, "
+                                << "request was made with downscale level " << downscaleLevel);
+    }
+
+    const std::lock_guard<std::mutex> lock(_mutex);
+
+    ALICEVISION_LOG_TRACE("[image] ImageCache: reading " << filename 
+                         << " with downscale level " << downscaleLevel
+                         << " from thread " << std::this_thread::get_id());
+
+    using TInfo = ColorTypeInfo<TPix>;
+
+    auto lastWriteTime = boost::filesystem::last_write_time(filename);
+    CacheKey keyReq(filename, TInfo::size, TInfo::typeDesc, downscaleLevel, lastWriteTime);
+
+    // find the requested image in the cached images
+    {
+        auto it = std::find(_keys.begin(), _keys.end(), keyReq);
+        if (it != _keys.end())
+        {
+            // image becomes LRU
+            _keys.erase(it);
+            _keys.push_back(keyReq);
+
+            _info.nbLoadFromCache++;
+
+            ALICEVISION_LOG_TRACE("[image] ImageCache: " << toString());
+            return _imagePtrs.at(keyReq).get<TPix>();
+        }
+    }
+
+    // retrieve image size
+    int width, height;
+    readImageSize(filename, width, height);
+    unsigned long long int memSize = (width / downscaleLevel) * (height / downscaleLevel) * sizeof(TPix);
+ 
+    // add image to cache if it fits in capacity
+    if (memSize + _info.contentSize <= _info.capacity) 
+    {
+        load<TPix>(keyReq);
+
+        ALICEVISION_LOG_TRACE("[image] ImageCache: " << toString());
+        return _imagePtrs.at(keyReq).get<TPix>();
+    }
+
+    // retrieve missing capacity
+    unsigned long long int missingCapacity = memSize + _info.contentSize - _info.capacity;
+
+    // find unused image with size bigger than missing capacity
+    // remove it and add image to cache
+    {
+        auto it = _keys.begin();
+        while (it != _keys.end())
+        {
+            const CacheKey& key = *it;
+            const CacheValue& value = _imagePtrs.at(key);
+            if (value.useCount() == 1 && value.memorySize() >= missingCapacity)
+            {
+                _info.nbImages--;
+                _info.contentSize -= value.memorySize();
+                _imagePtrs.erase(key);
+                _keys.erase(it);
+
+                _info.nbRemoveUnused++;
+
+                load<TPix>(keyReq);
+
+                ALICEVISION_LOG_TRACE("[image] ImageCache: " << toString());
+                return _imagePtrs.at(keyReq).get<TPix>();
+            }
+            ++it;
+        }
+    }
+
+    // remove as few unused images as possible
+    while (missingCapacity > 0)
+    {
+        auto it = std::find_if(_keys.begin(), _keys.end(), [this](const CacheKey& k){
+            return _imagePtrs.at(k).useCount() == 1;
+        });
+        
+        if (it != _keys.end())
```

Comment:
```
if(it == _keys.end())
{
    break;
}
```
Then the main code don't need to be in a sub-scope, which is better for readability.